crates/red_knot_python_semantic/resources/mdtest/generics/functions.md

```diff
@@ -162,61 +147,39 @@ parameters simultaneously.
 def two_params[T](x: T, y: T) -> T:
     return x
 
-# TODO: no error
 # TODO: revealed: str
-# error: [invalid-argument-type]
-# error: [invalid-argument-type]
-reveal_type(two_params("a", "b"))  # revealed: T
+reveal_type(two_params("a", "b"))  # revealed: Literal["a", "b"]
 
-# TODO: no error
 # TODO: revealed: str | int
-# error: [invalid-argument-type]
-# error: [invalid-argument-type]
-reveal_type(two_params("a", 1))  # revealed: T
+reveal_type(two_params("a", 1))  # revealed: Literal["a", 1]
 ```
 
 ```py
 def param_with_union[T](x: T | int, y: T) -> T:
     return y
 
-# TODO: no error
 # TODO: revealed: str
-# error: [invalid-argument-type]
-reveal_type(param_with_union(1, "a"))  # revealed: T
+reveal_type(param_with_union(1, "a"))  # revealed: Literal["a"]
```

Comment:
This case is a bit subtle, I had to actually try to write a body for `param_with_union` in which we could return `1` from this call, before realizing it's not possible because `T | int` is not assignable to `T`, so before we can return `x` we have to exclude `int` from its type, meaning this call can't return `1`.

Reply:
(Trying to write that function also highlighted a bug that currently we don't think `T & ~int` is assignable to `T`, but it should be.)

Reply:
I added some extra tests that exercise the logic I actually care about here (namely, that we try to infer the smallest specialization that satisfies all the constraints).  I went with a function with signature `(T | None) -> T`.  I think there's technically no way to correctly instantiate that signature, so for now I'm accepting the `[invalid-return-type]` error.  Once we support `list[T]`, I can make it return that.